o2a/mappers/java_mapper.py

```diff
@@ -0,0 +1,108 @@
+        props: PropertySet,
+        jar_files: List[str],
+        **kwargs,
+    ):
+        ActionMapper.__init__(
+            self, oozie_node=oozie_node, dag_name=dag_name, name=name, props=props, **kwargs
+        )
+        self.file_extractor = FileExtractor(oozie_node=oozie_node, props=self.props)
+        self.archive_extractor = ArchiveExtractor(oozie_node=oozie_node, props=self.props)
+        self.main_class: Optional[str] = None
+        self.java_opts: Optional[List[str]] = None
+        self.args: Optional[List[str]] = None
+        self.hdfs_files: Optional[List[str]] = None
+        self.hdfs_archives: Optional[List[str]] = None
+        self.prepare_extension: PrepareMapperExtension = PrepareMapperExtension(self)
+        self.jar_files: List[str] = jar_files if jar_files else []
+        self.jar_files_in_hdfs: List[str] = []
+        self._compute_jar_files_in_hdfs()
+
+    def on_parse_node(self):
+        super().on_parse_node()
+        _, self.hdfs_files = self.file_extractor.parse_node()
+        _, self.hdfs_archives = self.archive_extractor.parse_node()
+        self._extract_java_data()
+
+    def to_tasks_and_relations(self):
+        action_task = Task(
+            task_id=self.name,
+            template_name="java.tpl",
+            template_params=dict(
+                props=self.props,
+                hdfs_files=self.hdfs_files,
+                hdfs_archives=self.hdfs_archives,
+                main_class=self.main_class,
+                jar_files_in_hdfs=self.jar_files_in_hdfs,
+                args=self.args,
+            ),
+        )
+        tasks = [action_task]
+        relations: List[Relation] = []
+        prepare_task = self.prepare_extension.get_prepare_task()
+        if prepare_task:
+            tasks, relations = self.prepend_task(prepare_task, tasks, relations)
+        return tasks, relations
+
+    def required_imports(self) -> Set[str]:
+        return {"from airflow.utils import dates", "from airflow.contrib.operators import dataproc_operator"}
+
+    def _compute_jar_files_in_hdfs(self):
+        hdfs_app_prefix = self.props.job_properties["oozie.wf.application.path"]
+        for file in self.jar_files:
+            self.jar_files_in_hdfs.append(hdfs_app_prefix + "/" + LIB_FOLDER + "/" + file)
+
+    def _extract_java_data(self):
+        """Extracts Java node data."""
+        root = self.oozie_node
+        props = self.props
+        if "mapred.child.java.opts" in props.merged:
+            self.java_opts.extend(props.merged["mapred.child.java.opts"].split(" "))
+        if "mapreduce.map.java.opts" in props.merged:
+            self.java_opts.extend(props.merged["mapreduce.map.java.opts"].split(" "))
+        self.main_class = xml_utils.get_tag_el_text(root=root, tag="main-class", props=props)
+        java_opts_string = xml_utils.get_tag_el_text(root=root, tag="java-opts", props=props)
+        if java_opts_string:
+            self.java_opts = java_opts_string.split(" ")
```

Comment:
Aren't we overriding configuration set above by "mapred.child.java.opts" and "mapreduce.map.java.opts" here?
As I understand it we should preserve both.

From the [docs](https://oozie.apache.org/docs/5.1.0/WorkflowFunctionalSpec.html#a3.2.6_Java_Action):
> The java-opts and java-opt elements, if present, contains the command line parameters which are to be used to start the JVM that will execute the Java application. Using this element is equivalent to using the mapred.child.java.opts or mapreduce.map.java.opts configuration properties, with the advantage that Oozie will append to these properties instead of simply setting them (e.g. if you have one of these properties specified in mapred-site.xml, setting it again in the configuration element will override it, but **using java-opts or java-opt will instead append to it, preserving the original value**).

Reply:
Yep. True. I've done exactly this few lines above for java_opt but missed that one for java-opts. Shame on me.

Reply:
Fixed in fcffac2

Reply:
Thanks, kudos for the new test as well :+1: